added ddm_test_validate.py and logger

sdk-bonsai opened the pull request:
# DDM validation using simulator data

## Description

Modification adds a file ddm_test_validate.py. It addresses the problem of validating the trained ddm using the simulator, and logs two log files: 1) simulator state, action, config logs and 2) ddm state, action , config logs for same actions and configs. The logged states in the two csv files can be directly compared to see the difference and error propagation. No visualization or other changes are added there. The DDM_predictor file is also kept separate. 

The goal was to keep test and validation simplified and borrow same sim config as used in the brain training integration using microsoft-bonsai-api. So, functions are written only once and re-used from simulator-bonsai integration template as exists in mba samples. Further work or iteration could include connecting to bonsai-platform. But, test_random_policy function is sufficient for DDM explore experimentation, testing and validation. 

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Added logging facility to log both DDM prediction logs and Simulator logs for same actions and configs.
- [ ] Modified episode start to add symetric initial state and config randomization, and a function to accomodate both states/actions/configs with and without prefix. 
- [ ] Added moab sim folder inside a sim folder to allow using same functions as for platform integration.
- [ ] Added policies module from mba as a placeholder to test DDM using benchmark control methods for testing more planned test cases. This can have further development.
- [ ] ddm_test_validate.py will need more documentation to be added later.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Reviewer Tasks

List here all the tasks the reviewer needs to review:

- [ ] Test `examples/`


**Test Configuration**:
* OS:
* Python Version:
* Bonsai SDK Version:


## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
![image](https://user-images.githubusercontent.com/31707752/109709430-52674880-7b51-11eb-9a11-f545703472ce.png)

akzaidi commented:
@sdk-bonsai : thanks for the PR! Couple questions:

* there are four csv datasets, what are their differences?
* this repo is meant to just be the `ddm` library. Let's remove `policies.py` and entire `sim` folder
    - `sim/cartpole` - any differences from `main` branch in `microsoft-bonsai-api/Python/samples/cartpole`? If not, let's remove and just mention the commit SHA so users can revert if they want to replicate that specific environment. 
    - `sim/moab` - keep this in your branch in `microsoft-bonsai-api`

akzaidi commented:
I'm going to commit directly here to move everything from `ddm_test_validate` to `ddm_predictor`.

akzaidi commented:
verified with @sdk-bonsai in meeting, will try and document sims in the readme and remove from branch, and remove duplicated csv.

akzaidi commented on conf/config.yaml:
```diff
@@ -1,4 +1,4 @@
-  - data: cartpole_st1_at.yaml
-  - model: SVR.yaml
-  - simulator: gboost_cartpole.yaml
+  - data: moab_100epi.yaml
```
you haven't commited `moab_100epi.yaml`. Do note that you can also keep the defaults in `conf/config.yaml` and then use the CLI to pass new configs, e.g.: `python ddm_trainer.py data=moab_100epi`

sdk-bonsai replied:
I am removing the sim folder, and adding an entry point to `ddm_test_validate`.  Based on our discussion earlier today, we can separate the ddm_predictor into ddm_connector for bonsai platform integration and use `ddm_test_validate` to do random policy comparison with simulator at exactly same frequency. For this PR, I am just making the changes to ddm_test_validate: 1) removing sim,  2) adding entry point for user to include their sim locally, and   3) adding logging config to all simulator configs, and rest of comments below.

sdk-bonsai replied:
Updated these changes with recent commit.

akzaidi commented on conf/simulator/moab_simparam.yaml:
```diff
@@ -23,4 +22,5 @@ simulator:
     ]
-  policy: bonsai
+  policy: random
+  logging: enable
```
you've added `logging` here but not in the other `simulator/*.yaml` files. Still checking the rest of the files but we should keep the examples consistent.

akzaidi commented on ddm_test_validate.py:
```diff
@@ -0,0 +1,214 @@
+import datetime
+import pathlib
+import os
+import time
```
unused imports. we can probably remove?

sdk-bonsai replied:
Actually these are used for logging in line 49-62:
`if log_file=="enable":
            current_time = datetime.datetime.now().strftime("%Y-%m-%d-%H-%M-%S")
            log_file = current_time + "_" + env_name + "_log.csv"
            log_file2 = current_time + "_" + "SIM" + "_log.csv"
            logs_directory = pathlib.Path(log_file).parent.absolute()
            if not pathlib.Path(logs_directory).exists():
                print(
                    "Directory does not exist at {0}, creating now...".format(
                        str(logs_directory)
                    )
                )
                logs_directory.mkdir(parents=True, exist_ok=True)
        self.log_file = os.path.join(log_path, log_file)
        self.log_file2 = os.path.join(log_path, log_file2 `

sdk-bonsai replied:
Removed `time`, that was used during tests only, but not really necessary.

akzaidi commented on ddm_test_validate.py:
```diff
@@ -0,0 +1,214 @@
+import pathlib
+import os
+import time
+from distutils.util import strtobool
```
also unused

sdk-bonsai replied:
`strtobool` is unused, removing that one. The others are same as in above comment.

akzaidi commented on ddm_test_validate.py:
```diff
@@ -0,0 +1,214 @@
+
+import numpy as np
+import pandas as pd
+from azure.core.exceptions import HttpResponseError
+from dotenv import load_dotenv, set_key
+from microsoft_bonsai_api.simulator.client import BonsaiClient, BonsaiClientConfig
+from microsoft_bonsai_api.simulator.generated.models import (
+    SimulatorInterface,
+    SimulatorSessionResponse,
+    SimulatorState,
+)
+
```
likewise, you aren't connecting to bonsai in this script so you don't need these imports

sdk-bonsai replied:
Yes, I forgot to remove before. Will remove in the next commit.

akzaidi commented on ddm_test_validate.py:
```diff
@@ -0,0 +1,214 @@
+
+if __name__ == "__main__":
+
+    main()
```
hmm, it seems to me this entire script is basically the same as having the `policy=random` flag in `ddm_predictor`, although that script was defined for inferencing the ddm, not the original simulator. That said, it could be used for that purpose?

sdk-bonsai replied:
The intent is to allow multiple types of tests. Random policy is one of them. Coasting might be another. There might be other case specific tests for specific use case that might require to be scripted specifically. So, perhaps we could retain policies.py, and use that as a place for users to define their own test cases. (Initially, I wasn't in favor of policies.py. It seemed odd, but then when I see how much testing simulators is important, and how much in practice users ignore testing simulators (or ddms), it seems reasonable to keep it, now. We can always change it, or change its form once the software design for DDM firms up. I think that is going to be some incremental process.

akzaidi commented on environment.yml:
```diff
@@ -35,3 +35,5 @@ dependencies:
       - python-dotenv==0.15.0
       - hydra-core==1.0.5
       - natsort==7.1.0
+      - pyrr==0.10.3
+      - pyrsistent==0.16.0
```
these are the moab imports right?

sdk-bonsai replied:
Yes, Moab imports.  Will remove them from here. It is mostly for python simulators that facilitate matrix  and quaternion computations for robotics purposes. Such simulators will seldom have need of a ddm, we just testing it here.

akzaidi commented on ddm_test_validate.py:
```diff
@@ -0,0 +1,214 @@
+        self.config_keys = configs
+        self.state_keys = states
+        self.action_keys = actions
+        self.sim_physics = SimulatorSession()
```
ahhh okay, so you have both the original moab under `sim_physics` and the ddm model under `dd_model`. And you're running them both in the `test_random_policy` loop.

sdk-bonsai replied:
Yes, both run in the loop and get logged in separate files.

akzaidi commented on ddm_test_validate.py:
```diff
@@ -0,0 +1,214 @@
+    def episode_step(self, action: Dict[str, int]):
+
+        input_list = [
+            list(self.state.values()),
+            list(self.config.values()),
+            list(action.values()),
```
with numpy arrays it's much safter to do `np.ndarray.values().tolist()` rather than `list(np.ndarray.values())`. The former guarantees you get a list (potentially nested) of primitive python types back. the latter might give you `List[np.ndarray, np.ndarray]`  if you have more than 1 dimension for example, which isn't the case here but just in case...

sdk-bonsai replied:
That is a good idea, and I will use in other places. In this particular case, the states, config and action are `dict_values`, due to the structure of rest of the ddm implementation.

sdk-bonsai replied:
The state ,config and action are dict_values, so `.tolist` isn't available here. But, that is a good point, and I will use that in other places.

akzaidi reviewed (changes requested):
sorry this took so long to review fully. My thoughts:

1. I don't think we should keep the `sim` directory or `policies.py`, etc. those should remain in their original repos (or just a separate branch for moab in MBA to prevent redundancy against moabsim-py)
2. I see the utility in having both sims running with same initialization and at the same control frequency and logging together, but most of `ddm_test_validate` is duplicated of `ddm_predictor`. I think we can just add a function in `ddm_predictor` that initializes the original sim along with the DDM, and does the same as `ddm_test_validate.test_random_policy`
3. We can keep the logged data, and merge in the edits from the new `evaluate` methods to run evaluations after doing the additional sim + ddm log collection

sdk-bonsai commented:
@akzaidi , updated with a new commit. Following changes are included in the new commit: deleted sim and policies, added entry point for a simulator, test_policies function for allowing testing of various scenarios, and updates based on rest of the comments.

akzaidi commented:
@sdk-bonsai `offnames.csv` was renamed to `csv_data/moab_sim_5KepiPosVelMinStates.csv` , breaking our tests

```bash
================================================================ FAILURES =================================================================
_____________________________________________________________ test_diff_names _____________________________________________________________

    def test_diff_names():
    
        base_model = BaseModel()
        X, y = base_model.load_csv(
            dataset_path=os.path.join(data_dir, "off_names.csv"),
            input_cols=["x_position", "x_velocity", "angle_position", "angle_velocity",],
            output_cols=["angle_position", "angle_velocity"],
            augm_cols=["command", "length", "masspole"],
>           max_rows=1000,
        )

tests/test_loaders.py:82: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <base.BaseModel object at 0x7ffcc920c2d0>, dataset_path = 'csv_data/off_names.csv'
input_cols = ['x_position', 'x_velocity', 'angle_position', 'angle_velocity'], augm_cols = ['command', 'length', 'masspole']
output_cols = ['angle_position', 'angle_velocity'], iteration_order = -1, episode_col = 'episode', iteration_col = 'iteration'
drop_nulls = True, max_rows = 1000

    def load_csv(
        self,
        dataset_path: str,
        input_cols: Union[str, List[str]] = "state",
        augm_cols: Union[str, List[str]] = ["action_command"],
        output_cols: Union[str, List[str]] = "state",
        iteration_order: int = -1,
        episode_col: str = "episode",
        iteration_col: str = "iteration",
        drop_nulls: bool = True,
        max_rows: Union[int, None] = None,
    ) -> Tuple[np.ndarray, np.ndarray]:
        """Read CSV data into two datasets for modeling
    
        Parameters
        ----------
        dataset_path : str
            path to csv dataset
        input_cols : Union[str, List[str]], optional
            list of columns represent the inputs to the dynamical system in the raw dataset. Can either be a string which is then matched for all columns in the dataset, or a list of strings with exact matches, by default "state"
        augm_cols : Union[str, List[str]], optional
            Exact match of additional columns to use for modeling, such as the actions of the current iteration and any scenario/config parameters, by default ["action_command"]
        output_col : Union[str, List[str]], optional
            output columns of the dynamical system. Can either be a string which is then matched for any columns or a list of exact matches, by default "state"
        iteration_order : int, optional
            in the order of the raw dataset, what is the lag between iteration t and iteration t+1, by default -1
        max_rows : Union[int, None], optional
            max rows to read for a large dataset, by default None
    
        Returns
        -------
        Tuple[np.array, np.array]
            Features and labels for modeling
    
    
        Raises
        ------
        ValueError
            Data not found
        """
    
        csv_reader = CsvReader()
        if not os.path.exists(dataset_path):
>           raise ValueError(f"No data found at {dataset_path}")
E           ValueError: No data found at csv_data/off_names.csv

base.py:77: ValueError
```

is there a reason that file was renamed? has that file changed entirely?
